File: 144-binary-tree-preorder-traversal/binary-tree-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        # result = []

        # def dfs(node):
        #     if not node:
        #         return
        #     result.append(node.val)
        #     dfs(node.left)
        #     dfs(node.right)
        # dfs(root)
        # return result

        if not root:
            return []

        result = []
        stack = [root]

        while stack:
            node = stack.pop()
            result.append(node.val)
            if node.right:
                stack.append(node.right)
            if node.left:
                stack.append(node.left)
        return result

File: 144-binary-tree-preorder-traversal/binary-tree-preorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function(root) {
    /* const result = [];

    const helper = function(node) {
        if (node === null) return;
        result.push(node.val);
        helper(node.left);
        helper(node.right);
    }
    helper(root);
    return result; */

    if (root === null) return [];
    const stack = [root];
    const result = [];

    while (stack.length !== 0) {
        let node = stack.pop()
        result.push(node.val);
        if (node.right) {
            stack.push(node.right);
        }
        if (node.left) {
            stack.push(node.left)
        }

    }
    return result;

};